AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,276 @@
+{
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusReadyToPlay) {
+    if ([self.subnodes containsObject:_spinner]) {
+      [_spinner removeFromSupernode];
+      _spinner = nil;
+    }
+  }
+  
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusFailed) {
+    
+  }
+}
+
+- (void)restartVideo:(NSNotification *)notification
+{
+  if ( [[[notification object] asset] isEqual:_asset]) {
+    [[((AVPlayerLayer *)_playerNode.layer) player] seekToTime:CMTimeMakeWithSeconds(0, 1)];
+    
+    if (_autorepeat) {
+      [self play];
+    } else {
+      [self pause];
+    }
+  }
+}
+
+- (void)layoutDidFinish
+{
+  _playerNode.frame = self.bounds;
+}
+
+- (void)didLoad {
+  [super didLoad];
+  
+  UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapped)];
+  [self.view addGestureRecognizer:tap];
+}
+
+- (void)tapped
+{
+  if (_shouldBePlaying) {
+    [self pause];
+  } else {
+    [self play];
+  }
+}
+
+- (instancetype)initWithViewBlock:(ASDisplayNodeViewBlock)viewBlock didLoadBlock:(ASDisplayNodeDidLoadBlock)didLoadBlock
+{
+  ASDisplayNodeAssertNotSupported();
+  return nil;
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+
+  @try {
+    [_currentItem removeObserver:self forKeyPath:NSStringFromSelector(@selector(status))];
+  }
+  @catch (NSException * __unused exception) {
+    NSLog(@"unnecessary removal in fetch data");
+  }
+
+  {
+    ASDN::MutexLocker l(_lock);
+    _currentItem = [[AVPlayerItem alloc] initWithAsset:_asset];
+    [_currentItem addObserver:self forKeyPath:NSStringFromSelector(@selector(status)) options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew context:NULL];
+
+    if (((AVPlayerLayer *)_playerNode.layer).player) {
+      [((AVPlayerLayer *)_playerNode.layer).player replaceCurrentItemWithPlayerItem:_currentItem];
+    } else {
+      ((AVPlayerLayer *)_playerNode.layer).player = [[AVPlayer alloc] initWithPlayerItem:_currentItem];
+    }
+
+  }
+  if (_shouldAutoPlay) {
+    [self play];
+  }
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  
+  {
+    ASDN::MutexLocker l(_lock);
+    ((AVPlayerLayer *)_playerNode.layer).player = nil;
+    _shouldBePlaying = NO;
+  }
+}
+
+- (void)setPlayButton:(ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  _playButton = playButton;
+  
+  [self addSubnode:playButton];
+  
+  [_playButton addTarget:self action:@selector(play) forControlEvents:ASControlNodeEventTouchUpInside];
+}
+
+- (ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return _playButton;
+}
+
+- (void)setAsset:(AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if (ASObjectIsEqual(((AVURLAsset *)asset).URL, ((AVURLAsset *)_asset).URL)) {
+    return;
+  }
+  
+  _asset = asset;
+  
+  if (self.interfaceState & ASInterfaceStateFetchData) {
```

Comment:
This is as much a reminder for me as anything, but add:

// FIXME: Adopt -setNeedsFetchData when it is available